clash-lib/src/Clash/Core/Evaluator/Models.hs

```diff
@@ -0,0 +1,295 @@
+-- as the 'AsTerm' class (which converts back to Term) has instances for both.
+--
+data Evaluator = Evaluator
+  { evaluateWhnf :: Term -> State Env Value
+  , quote        :: Value -> State Env Nf
+  }
+
+-- | Evaluate a term to normal form without stopping at WHNF.
+--
+evaluateNf :: Evaluator -> Term -> State Env Nf
+evaluateNf (Evaluator e q) = e >=> q
+
+-- Local bindings are not stored in a VarEnv, as we still want to be
+-- able to access the keys (VarEnv simply indexes by unique).
+
+type EnvTyMap = Map TyVar Type
+type EnvTmMap = Map Id (Either Term Value)
+
+-- | EnvGlobals refers to global bindings in scope during evalaution, i.e.
+-- other top-level functions used in a program.
+--
+type EnvGlobals = VarEnv (Binding (Either Term Value))
+
+-- | EnvPrimsIO refers to the result of IO operations performed in primitives
+-- during evalaution e.g. creating a new ByteArray in the GHC frontend.
+--
+type EnvPrimsIO = (IntMap Value, Int)
+
+-- | An Environment contains all in scope terms and types while evaluating.
+-- This consists of
+--
+--   * local types and terms from the term being evaluated
+--   * top-level definitions in scope (i.e. global bindings)
+--   * IO results of primitive operations
+--   * in scope uniques and a means of generating uniques
+--
+-- When running the partial evaluator at different times, it may be desirable
+-- to keep the globals (some of which may have been partially evaluated) and
+-- the results of primitives in IO for future runs.
+--
+data Env = Env
+  { envTypes   :: !EnvTyMap
+  , envLocals  :: !EnvTmMap
+  , envGlobals :: !EnvGlobals
+  , envPrimsIO :: !EnvPrimsIO
+  , envTcMap   :: TyConMap
+  , envInScope :: InScopeSet
+  , envSupply  :: Supply
+  }
+
+instance Show Env where
+  show env = show (envTypes env, envLocals env, envPrimsIO env)
+
+-- This instance cannot be derived with Generic, as neither InScopeSet
+-- or Supply have instances for NFData. Instead, we use this instance
+-- which forces all fields which can be forced.
+--
+instance NFData Env where
+  rnf env = rnf (envTypes env)
+    `seq` rnf (envLocals env)
+    `seq` rnf (envGlobals env)
+    `seq` rnf (envPrimsIO env)
+
+mkEnv :: BindingMap -> EnvPrimsIO -> TyConMap -> InScopeSet -> Supply -> Env
+mkEnv bs = Env mempty mempty gs
+ where
+  gs = fmap Left <$> bs
+
+-- | Add a new term / value to the environment. The unique for the new
+-- element is guaranteed to be unique within the environment.
+--
+-- For new global bindings, we assume that it is possible to inline
+-- them later during evaluation.
+--
+insertEnv :: IdScope -> Id -> Either Term Value -> Env -> Env
+insertEnv scope i etv env =
+  case scope of
+    LocalId -> env
+      { envLocals = Map.insert i' etv (envLocals env)
+      , envInScope = extendInScopeSet (envInScope env) i'
+      }
+
+    GlobalId ->
+      let b = Binding i' noSrcSpan Inline etv
+       in env { envGlobals = extendVarEnv i' b (envGlobals env) }
```

Comment:
doesn't the `envInScope` only contain local ids? if so, then this can accidentally overwrite global binders. So perhaps find a different way to generate a unique global id.

Reply:
The `InScopeSet` is given as an argument to the evaluator (I think from normalize or rewrite state). So I'm genuinely not sure if it contains the global bindings or not...

Reply:
Then it won't

Reply:
But if this `insertEnv` is to be seen an "update action" for globals, then you can just `extendVarEnv` at `i`.